numpy/testing/utils.py

```diff
@@ -1718,3 +1718,72 @@ def tempdir(*args, **kwargs):
     tmpdir = mkdtemp(*args, **kwargs)
     yield tmpdir
     shutil.rmtree(tmpdir)
+
+
+class clear_and_catch_warnings(warnings.catch_warnings):
+    """ ``catch_warnings`` context manager that resets warning registry
+
+    Warnings can be slippery, because, whenever a warning is triggered, Python
+    adds a ``__warningregistry__`` member to the *calling* module.  This makes
+    it impossible to retrigger the warning in this module, whatever you put in
+    the warnings filters.  This context manager accepts a sequence of `modules`
+    as a keyword argument to its constructor and:
+
+    * stores and removes any ``__warningregistry__`` entries in given `modules`
+      on entry;
+    * resets ``__warningregistry__`` to its previous state on exit.
+
+    This makes it possible to trigger any warning afresh inside the context
+    manager without disturbing the state of warnings outside.
+    """
+    class_modules = ()
+
+    def __init__(self, record=False, module=None, modules=()):
+        """ Specify whether to record warnings, and modules to reset
+
+        For compatibility with Python 3.0, please consider all arguments to be
+        keyword-only.
+
+        Parameters
+        ----------
+        record : bool, optional
+            Specifies whether warnings should be captured by a custom
+            implementation of ``warnings.showwarning()`` and be appended to a
+            list returned by the context manager. Otherwise None is returned by
+            the context manager. The objects appended to the list are arguments
+            whose attributes mirror the arguments to ``showwarning()``.
+        module : None or module instance, optional
```

Comment:
This would be better as the last argument, as it will hardly ever be used. Could also be omitted, I don't think we will use it.